File: internal/cli/ctl/html.go
/*
Maddy Mail Server - Composable all-in-one email server.
Copyright © 2019-2020 Max Mazurov <fox.cpp@disroot.org>, Maddy Mail Server contributors

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

package ctl

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	maddycli "github.com/themadorg/madmail/internal/cli"
	"github.com/themadorg/madmail/internal/endpoint/chatmail"
	"github.com/urfave/cli/v2"
)

func init() {
	maddycli.AddSubcommand(&cli.Command{
		Name:        "html-export",
		Usage:       "Export default HTML files to a directory",
		ArgsUsage:   "DEST_DIR",
		Description: "Exports all embedded HTML, CSS, and SVG files from the chatmail endpoint to the specified directory.",
		Action: func(ctx *cli.Context) error {
			if ctx.NArg() < 1 {
				return cli.ShowCommandHelp(ctx, "html-export")
			}
			return htmlExport(ctx.Args().First())
		},
	})
	maddycli.AddSubcommand(&cli.Command{
		Name:        "html-serve",
		Usage:       "Configure maddy to serve HTML from a directory",
		ArgsUsage:   "WWW_DIR",
		Description: "Updates maddy.conf to serve chatmail HTML files from the specified external directory instead of embedded ones.",
		Action: func(ctx *cli.Context) error {
			if ctx.NArg() < 1 {
				return cli.ShowCommandHelp(ctx, "html-serve")
			}
			return htmlServe(ctx)
		},
	})
}

func htmlExport(destDir string) error {
	if err := os.MkdirAll(destDir, 0755); err != nil {
		return fmt.Errorf("failed to create directory: %v", err)
	}

	entries, err := chatmail.WWWFiles.ReadDir("www")
	if err != nil {
		return fmt.Errorf("failed to read embedded files: %v", err)
	}

	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		data, err := chatmail.WWWFiles.ReadFile("www/" + entry.Name())
		if err != nil {
			return fmt.Errorf("failed to read %s: %v", entry.Name(), err)
		}
		if err := os.WriteFile(filepath.Join(destDir, entry.Name()), data, 0644); err != nil {
			return fmt.Errorf("failed to write %s: %v", entry.Name(), err)
		}
		fmt.Printf("Exported: %s\n", entry.Name())
	}
	fmt.Printf("Successfully exported all files to %s\n", destDir)
	return nil
}

func htmlServe(ctx *cli.Context) error {
	wwwDir := ctx.Args().First()
	isEmbedded := wwwDir == "embedded" || wwwDir == "embed" || wwwDir == "internal"

	var absWWWDir string
	var err error
	if !isEmbedded {
		absWWWDir, err = filepath.Abs(wwwDir)
		if err != nil {
			return fmt.Errorf("failed to get absolute path: %v", err)
		}
	}

	cfgPath := ctx.String("config")
	if cfgPath == "" {
		return cli.Exit("Error: config is required", 2)
	}

	data, err := os.ReadFile(cfgPath)
	if err != nil {
		return fmt.Errorf("failed to read config: %v", err)
	}

	lines := strings.Split(string(data), "\n")
	var newLines []string
	updated := false
	inChatmail := false

	for i := 0; i < len(lines); i++ {
		line := lines[i]
		trimmed := strings.TrimSpace(line)

		// Detection of chatmail block start
		if strings.HasPrefix(trimmed, "chatmail ") && strings.HasSuffix(trimmed, "{") {
			inChatmail = true
			newLines = append(newLines, line)
			continue
		}

		if inChatmail && trimmed == "}" {
			if isEmbedded {
				// We don't add anything for embedded, and we should have filtered out www_dir
				updated = true
			} else {
				// Check if we already have www_dir in this block
				hasWWWDir := false
				for j := len(newLines) - 1; j >= 0; j-- {
					if strings.HasPrefix(strings.TrimSpace(newLines[j]), "chatmail ") {
						break
					}
					if strings.HasPrefix(strings.TrimSpace(newLines[j]), "www_dir ") {
						newLines[j] = "    www_dir " + absWWWDir
						hasWWWDir = true
						updated = true
						break
					}
				}
				if !hasWWWDir {
					newLines = append(newLines, "    www_dir "+absWWWDir)
					updated = true
				}
			}
			inChatmail = false
		}

		// Filter out existing www_dir if we are in chatmail block
		if inChatmail && strings.HasPrefix(trimmed, "www_dir ") {
			if isEmbedded {
				// Just omit it to "remove" it
				continue
			}
			// When updating, we replace it above at the end of the block or find it.
			// Actually, the logic above for updating is slightly flawed if we append it here too.
			// Let's just skip it here and the logic at '}' will handle re-insertion if needed.
			continue
		}

		newLines = append(newLines, line)
	}

	if !updated {
		return cli.Exit("Error: No chatmail block found in config or failed to update.", 2)
	}

	if err := os.WriteFile(cfgPath, []byte(strings.Join(newLines, "\n")), 0644); err != nil {
		return fmt.Errorf("failed to write config: %v", err)
	}

	if isEmbedded {
		fmt.Printf("Successfully updated %s to use embedded HTML files.\n", cfgPath)
	} else {
		fmt.Printf("Successfully updated %s to serve HTML from %s\n", cfgPath, absWWWDir)
		fmt.Println("\n⚠️  Note: Ensure the 'maddy' user has read access to this directory.")
		fmt.Println("If the directory is inside a home folder, you may need to move it to /var/lib/maddy/www.")
		fmt.Println("Example: sudo mv " + wwwDir + " /var/lib/maddy/www && sudo chown -R maddy:maddy /var/lib/maddy/www")
	}
	fmt.Println("\nPlease restart maddy to apply changes: sudo systemctl restart maddy")
	return nil
}

File: crates/chatmail/src/ctl/html.rs
// Copyright (C) 2026 themadorg
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! `chatmail html-export` / `html-serve` / `html-migrate` (Madmail `ctl/html.go`).

use std::io::IsTerminal;
use std::path::Path;

use chatmail_config::{load_config, update_config_www_dir, Args};
use chatmail_types::{ChatmailError, Result};
use chatmail_www::{export_www_files, migrate_www_dir};

use super::context::CtlContext;
use super::output::CtlOut;
use super::util::confirm;

pub async fn html_export(args: &Args, dest: &Path) -> Result<()> {
    let _ = CtlContext::from_args(args)?;
    let out = CtlOut::from_args(args, "html-export");
    let n = export_www_files(dest)?;
    out.done_msg(
        format!("Successfully exported {n} files to {}", dest.display()),
        serde_json::json!({ "dest": dest.display().to_string(), "files": n }),
        format!("Exported {n} files"),
    )
}

pub async fn html_serve(args: &Args, www_dir: &str) -> Result<()> {
    let _ctx = CtlContext::from_args(args)?;
    let out = CtlOut::from_args(args, "html-serve");

    let embedded = matches!(
        www_dir.trim().to_ascii_lowercase().as_str(),
        "embedded" | "embed" | "internal"
    );

    let www_path = if embedded {
        None
    } else {
        let p = Path::new(www_dir);
        if !p.is_dir() {
            return Err(ChatmailError::config(format!(
                "directory not found: {}",
                p.display()
            )));
        }
        Some(p.canonicalize().unwrap_or_else(|_| p.to_path_buf()))
    };

    if !args.config.is_file() {
        return Err(ChatmailError::config(format!(
            "config file not found: {} — pass --config",
            args.config.display()
        )));
    }

    update_config_www_dir(&args.config, www_path.as_deref())?;

    if out.is_json() {
        return out.done_msg(
            "",
            serde_json::json!({
                "config": args.config.display().to_string(),
                "embedded": embedded,
                "www_dir": www_path.as_ref().map(|p| p.display().to_string()),
            }),
            if embedded {
                "Updated config to use embedded HTML".into()
            } else {
                format!(
                    "Updated config to serve HTML from {}",
                    www_path.as_ref().unwrap().display()
                )
            },
        );
    }

    if embedded {
        out.line(format!(
            "Successfully updated {} to use embedded HTML files.",
            args.config.display()
        ));
    } else {
        let p = www_path.as_ref().unwrap();
        out.line(format!(
            "Successfully updated {} to serve HTML from {}",
            args.config.display(),
            p.display()
        ));
        out.blank();
        out.line("Ensure the chatmail service user can read this directory.");
        out.line(format!(
            "Example: sudo chown -R chatmail:chatmail {}",
            p.display()
        ));
    }
    out.blank();
    out.line("Restart chatmail to apply: sudo systemctl restart madmail");
    Ok(())
}

/// Convert custom `www_dir` Go templates to Minijinja and fix legacy QR/static assets.
///
/// Used interactively by operators and non-interactively from `madmail update`
/// (re-exec of the new binary after replace).
pub async fn html_migrate(args: &Args, yes: bool) -> Result<()> {
    let _ = CtlContext::from_args(args)?;
    let out = CtlOut::from_args(args, "html-migrate");

    if !args.config.is_file() {
        return Err(ChatmailError::config(format!(
            "config file not found: {} — pass --config",
            args.config.display()
        )));
    }

    let cfg = load_config(&args.config)?;
    let Some(www_dir) = cfg.www_dir.as_ref() else {
        let msg = "No custom www_dir in config (embedded site); nothing to migrate.";
        return out.done_msg(
            msg,
            serde_json::json!({
                "config": args.config.display().to_string(),
                "www_dir": null,
                "action": "noop_embedded",
            }),
            msg,
        );
    };

    if !www_dir.is_dir() {
        return Err(ChatmailError::config(format!(
            "www_dir is set but not a directory: {}",
            www_dir.display()
        )));
    }

    let dry = migrate_www_dir(www_dir, false)?;
    print_literal_brace_warnings(&out, &dry.literal_brace_warnings);

    if !dry.dry_needs_apply() {
        let msg = format!(
            "Custom www_dir {} has no Go-style HTML templates and no legacy /qr?data= QR usage \
             (already Minijinja / client-side QR, or no templates).",
            www_dir.display()
        );
        if !out.is_json() && !dry.literal_brace_warnings.is_empty() {
            out.blank();
            out.line(
                "Literal `{%` / `{{` warnings above are not auto-fixed — wrap with \
                 `{% raw %}…{% endraw %}` if they are plain text or URLs.",
            );
        }
        return out.done_msg(
            &msg,
            serde_json::json!({
                "config": args.config.display().to_string(),
                "www_dir": www_dir.display().to_string(),
                "scanned": dry.scanned,
                "go_style_files": dry.go_style_files,
                "qr_legacy_files": dry.qr_legacy_files,
                "literal_brace_warnings": dry.literal_brace_warnings,
                "action": "noop_already_migrated",
            }),
            "No Go-style templates or legacy QR found",
        );
    }

    if !out.is_json() {
        if !dry.go_style_files.is_empty() {
            out.line(format!(
                "Found {} Go-style HTML template(s) under {}:",
                dry.go_style_files.len(),
                www_dir.display()
            ));
            for (i, f) in dry.go_style_files.iter().enumerate() {
                if i >= 12 {
                    out.line(format!(
                        "  … and {} more",
                        dry.go_style_files.len().saturating_sub(12)
                    ));
                    break;
                }
                out.line(format!("  - {f}"));
            }
            out.blank();
        }
        if !dry.qr_legacy_files.is_empty() {
            out.line(format!(
                "Found {} legacy QR / client-QR asset issue(s) under {}:",
                dry.qr_legacy_files.len(),
                www_dir.display()
            ));
            for (i, f) in dry.qr_legacy_files.iter().enumerate() {
                if i >= 12 {
                    out.line(format!(
                        "  … and {} more",
                        dry.qr_legacy_files.len().saturating_sub(12)
                    ));
                    break;
                }
                out.line(format!("  - {f}"));
            }
            out.blank();
        }
        out.line(
            "This rewrites Go html/template → Minijinja and/or `/qr?data=` → client-side QR \
             (setQrCodeImage + qrcode.min.js). Backups: *.go-template.bak / main.js.qr-compat.bak.",
        );
    }

    let interactive = std::io::stdin().is_terminal() && !out.is_json();
    let should_apply = if yes {
        true
    } else if interactive {
        confirm(
            "Migrate custom www (Go templates → Minijinja and/or legacy QR → client-side)?",
            false,
        )?
    } else {
        if !out.is_json() {
            out.line(
                "Non-interactive session: not converting. Re-run with: \
                 madmail html-migrate --yes",
            );
        }
        return out.done_msg(
            "Skipped (non-interactive; pass --yes to convert)",
            serde_json::json!({
                "config": args.config.display().to_string(),
                "www_dir": www_dir.display().to_string(),
                "scanned": dry.scanned,
                "go_style_files": dry.go_style_files,
                "qr_legacy_files": dry.qr_legacy_files,
                "literal_brace_warnings": dry.literal_brace_warnings,
                "action": "skipped_noninteractive",
            }),
            "Skipped non-interactive",
        );
    };

    if !should_apply {
        return out.done_msg(
            "Left custom www templates and QR assets unchanged.",
            serde_json::json!({
                "config": args.config.display().to_string(),
                "www_dir": www_dir.display().to_string(),
                "scanned": dry.scanned,
                "go_style_files": dry.go_style_files,
                "qr_legacy_files": dry.qr_legacy_files,
                "literal_brace_warnings": dry.literal_brace_warnings,
                "action": "declined",
            }),
            "Declined",
        );
    }

    let report = migrate_www_dir(www_dir, true)?;
    if !out.is_json() {
        if !report.migrated.is_empty() || !report.qr_migrated.is_empty() {
            out.line(format!(
                "Migrated {} HTML file(s); QR/static updates: {}; backups: {}",
                report.migrated.len(),
                report.qr_migrated.len(),
                report.backups.len()
            ));
            for f in &report.migrated {
                out.line(format!("  ✓ {f}"));
            }
            for f in &report.qr_migrated {
                if !report.migrated.contains(f) {
                    out.line(format!("  ✓ QR/static: {f}"));
                }
            }
            if report.qrcode_js_copied {
                out.line("  ✓ copied qrcode.min.js from embedded assets");
            }
            if report.main_js_qr_helper_added {
                out.line("  ✓ appended setQrCodeImage to main.js");
            }
        }
        if !report.errors.is_empty() {
            for e in &report.errors {
                out.line(format!("  ⚠ {e}"));
            }
        }
        print_literal_brace_warnings(&out, &report.literal_brace_warnings);
        if !report.literal_brace_warnings.is_empty() {
            out.blank();
            out.line(
                "Literal `{%` / `{{` warnings are not auto-fixed — wrap with \
                 `{% raw %}…{% endraw %}` if they are plain text or URLs.",
            );
        }
    }

    let action = if report.migrated.is_empty()
        && report.qr_migrated.is_empty()
        && !report.qrcode_js_copied
        && !report.main_js_qr_helper_added
    {
        "noop_already_migrated"
    } else {
        "migrated"
    };

    out.done_msg(
        format!(
            "Migrated {} template(s), {} QR/static update(s) under {}",
            report.migrated.len(),
            report.qr_migrated.len(),
            www_dir.display()
        ),
        serde_json::json!({
            "config": args.config.display().to_string(),
            "www_dir": www_dir.display().to_string(),
            "scanned": report.scanned,
            "go_style_files": report.go_style_files,
            "migrated": report.migrated,
            "qr_legacy_files": report.qr_legacy_files,
            "qr_migrated": report.qr_migrated,
            "qrcode_js_copied": report.qrcode_js_copied,
            "main_js_qr_helper_added": report.main_js_qr_helper_added,
            "literal_brace_warnings": report.literal_brace_warnings,
            "backups": report.backups,
            "errors": report.errors,
            "action": action,
        }),
        format!(
            "Migrated {} file(s), {} QR/static",
            report.migrated.len(),
            report.qr_migrated.len()
        ),
    )
}

fn print_literal_brace_warnings(out: &CtlOut, warnings: &[String]) {
    if out.is_json() || warnings.is_empty() {
        return;
    }
    out.blank();
    out.line(format!(
        "⚠ {} possible literal `{{%` / `{{{{` sequence(s) (Minijinja may 500):",
        warnings.len()
    ));
    for (i, w) in warnings.iter().enumerate() {
        if i >= 8 {
            out.line(format!("  … and {} more", warnings.len().saturating_sub(8)));
            break;
        }
        out.line(format!("  - {w}"));
    }
}
